### ISSUE TITLE
In "Project > Properties: Resource" "Restore Defaults" removes explicit project encoding

### ISSUE THREAD

**Issue by howlger** [eclipse-platform/eclipse.platform.resources#152]
In Eclipse 4.24RC2a, in _Project > Properties: Resource_, hitting _Restore Defaults_ and _Apply and Close_ removes the project's UTF-8 default encoding that was introduced in Eclipse 4.24 by [bug 479450](https://bugs.eclipse.org/bugs/show_bug.cgi?id=479450).

Expected behavior: in _Project > Properties: Resource_, _Restore Defaults_ should set the encoding of the project explicitly to UTF-8.

**Comment by iloveeclipse** [comment]
> Expected behavior: in _Project > Properties: Resource_, _Restore Defaults_ should set the encoding of the project explicitly to UTF-8.

This is probably wrong assumption, see below.

The "default" encoding for a project is workspace encoding.
The "default" encoding for a workspace is system encoding.

Bug 479450 sets *explicit* encoding for *new* projects and workspaces to UTF-8 because we *think* it is what most appropriate for new projects, but it is probably not "default" for people that write in Japanese on a Windows system (no idea which encoding is used there, but I assume not utf-8).

But I see, that this is surprising *at least*.

**Comment by laeubi** [comment]
Maybe we can store a properties file in the bundles persistent area with the "inital defaults" when a workspace is first created and then later on workspace init set this as the new preference defaults (if given). That way 'old' workspace won't be changed and even if preferences are removed we always know the "initial" state.

**Comment by howlger** [comment]
Storing the properties file in the bundles persistent area, would result in a different behavior of _Restore Defaults_ when sharing the project. When a project is shared without its `.settings` folder (which many sadly do because they read it on Stack Overflow), it is impossible to determine whether the project was created by an older Eclipse without an explicit encoding or by a newer Eclipse with an explicit UTF-8 encoding that was removed later.

Since the workspace encoding is now UTF-8 by default, upgrading to Eclipse 2022-06 (4.24) will change the encoding of projects without explicit encoding to UTF-8 (inherited from the workspace) and a _"no explicit encoding set"_ will be shown with a quick fix to set an explicit UTF-8 project encoding. There is no second quick fix for backwards compatibility yet to set an encoding that is derived from the system which would give the user the choice.

Whatever, for _Restore Defaults_ we cannot not ask the user. We have to decide whether _Restore Defaults_ removes an explicit project encoding for reasons of backwards compatibility, or sets an explicit UTF-8 encoding like it is done when creating a new project. I would prefer _Restore Defaults_ sets an explicit UTF-8 encoding for the following reasons:

- Backwards compatibility is already broken by the default UTF-8 workspace encoding
- Consistency to create a new project
- Consistency to the quick fix
- Avoiding of the _"no explicit encoding set"_ warning

I reported the issue because the current behavior does not meet my expectation shortly after creating a new project. I think _Project > Properties: Resource: Restore Defaults_ is not that important. I guess a second quick fix to set an explicit project encoding that is derived from the system would be more helpful. What do you think?

**Comment by iloveeclipse** [comment]
> Backwards compatibility is already broken by the default UTF-8 workspace encoding

This is not true. Existing workspaces aren't touched, projects imported get a warning, so "breaking" something is up to user who ignores that warning.

> I think Project > Properties: Resource: Restore Defaults is not that important.

Why?

>  I guess a second quick fix to set an explicit project encoding that is derived from the system would be more helpful. 

So is that more *important*, or something else?

> Whatever, for Restore Defaults we cannot not ask the user.

This is interesting idea. Why not actually. 
- So what would we like to ask?
- Something like: "Should we reset to default system encoding (XYZ) or to UTF-8 (recommended)"
- If the system encoding is UTF-8, no question or also a question? 
- Something like "Project/Workspace already uses recommended UTF-8 encoding. Reset to system default (XYZ)?

**Comment by laeubi** [comment]
> Storing the properties file in the bundles persistent area, would result in a different behavior of _Restore Defaults_ when sharing the project.

Wouldn't it always be the case if I import a project in a workspace without setting the projects encoding? I probably don't understand the case then, I assumed it was that the "always use utf8 fro new workspaces" was lost when resetting to defaults because from that on we don't know it is a 'new workspace with utf-8 default' ... ?!?

**Comment by howlger** [comment]
When I do the following on Windows 10, the project encoding gets changed from Cp1252 to UTF-8 (causing a _"Syntax error on token "Invalid Character", delete this token"_ error in the Java file for the class name):

1. With _Eclipse IDE for Java Developers **2022-03**_:
    1. Create a new **workspace**<br>→ preferences _General > Workspace: Text file encoding: Default (**Cp1252**)_
    2. Inside the new workspace create a new **project** _Hello Wörld_<br>→ _Project > Properties: Resource: Text file encoding: Inherited from container (**Cp1252**)_
    3. Inside the new project create a new **class** `Äpp`<br>→ right-click `Äpp.java`: _Properties: Resource: Text file encoding: Default (inherited from container: **Cp1252**)_
2. With _Eclipse IDE for Java Developers **2022-06** RC1_ open the workspace<br>→ preferences _General > Workspace: Text file encoding: Default (**UTF-8**)_<br>→ _Project > Properties: Resource: Text file encoding: Inherited from container (**UTF-8**)_<br>→ right-click `Äpp.java`: _Properties: Resource: Text file encoding: Default (inherited from container: **UTF-8**)_<br>→ _"Project 'Hello Wörld' has no explicit encoding set"_ warning

Is this the expected behavior?

I would prefer that _Project > Properties: Resource: Restore Defaults_ sets the explicit UTF-8 project encoding without any prompt (in general, I don't expect to be asked by _Restore Defaults_). But if there would be two quick fixes to set the project encoding to UTF-8 and to the encoding derived from the system respectively, removing the explicit project encoding would make more sense to me. So they are different things, but somehow related.

No explicit project encoding can cause issues independent of the Eclipse version used. That's why I think the new warning is a good idea.

**Comment by iloveeclipse** [comment]
> Is this the expected behavior?

No. I assume you use oomph or something else that sets the explicit workspace encoding automatically. I can't reproduce that behavior with plain 4.23 / 4.24 SDK builds. Might be your package set some value in the eclipse.ini. if so, that's a bug because it forces existing workspaces to use different encoding.

> I would prefer that Project > Properties: Resource: Restore Defaults sets the explicit UTF-8 project encoding without any prompt

I would prefer that too, but this is again something we can't know **for sure** if that is safe or not. So with your example, previously Cp1252 encoded data will be broken after that. On the other side, one can argue, it iwas a user decision to change that value to something else, so user should know about consequences.

> No explicit project encoding can cause issues independent of the Eclipse version used

Sure.

**Comment by iloveeclipse** [comment]
Hmm. The eclipse.ini seem to be OK. So it can be either some bundle that does something either on startup or via p2.inf file, or Java 17 which is bundled into. I test on RHEL 7 with Java 11.

**Comment by jonahgraham** [comment]
> I just unzipped [`eclipse-java-2022-03-R-win32-x86_64.zip`](https://www.eclipse.org/downloads/download.php?file=/technology/epp/downloads/release/2022-03/R/eclipse-java-2022-03-R-win32-x86_64.zip) and [`eclipse-java-2022-06-R-win32-x86_64.zip`](https://download.eclipse.org/technology/epp/staging/eclipse-java-2022-06-R-win32-x86_64.zip). Could somebody else please check if this is reproducible? Maybe it is an issue of the IDE packages, but not of the SDK.

Hi @howlger - thanks for [reaching out](https://twitter.com/howlger/status/1534906834712240131) to me on Twitter.

I cannot reproduce what you saw in https://github.com/eclipse-platform/eclipse.platform.resources/issues/152#issuecomment-1151113878 - on my Windows machine when I open the workspace initially created with 2022-03 using 2022-06 IDE the encoding stays as default, cp1252. When I create a new workspace with 2022-06 IDE the encoding is UTF-8.

Additionally, in the workspace I initially created with 2022-03, if I create a project with 2022-06 the encoding of that project is Cp1252 and that encoding is stored in `.settings/org.eclipse.core.resources.prefs` as `encoding/<project>=Cp1252` and consequently that project has no warning.

I have two differences in setup - Windows 11 and I am using 2022-06 RC2 (which will be available for testing officially later today).

To be explicit, this is what I see:

When I do the following on **Windows 11**, the project encoding does not change. 

1. With _Eclipse IDE for Java Developers **2022-03**_:
    1. Create a new **workspace**<br>→ preferences _General > Workspace: Text file encoding: Default (**Cp1252**)_
    2. Inside the new workspace create a new **project** _Hello Wörld_<br>→ _Project > Properties: Resource: Text file encoding: Inherited from container (**Cp1252**)_
    3. Inside the new project create a new **class** `Äpp`<br>→ right-click `Äpp.java`: _Properties: Resource: Text file encoding: Default (inherited from container: **Cp1252**)_
2. With _Eclipse IDE for Java Developers **2022-06** RC2_ open the workspace<br>→ preferences _General > Workspace: Text file encoding: Default (**Cp1252**)_<br>→ _Project > Properties: Resource: Text file encoding: Inherited from container (**Cp1252**)_<br>→ right-click `Äpp.java`: _Properties: Resource: Text file encoding: Default (inherited from container: **Cp1252**)_<br>→ _"Project 'Hello Wörld' has no explicit encoding set"_ warning
3. In the same workspace as above, with **2022-06**
    1. Inside the workspace create a new **project** _Hello Wörld2_<br>→ _Project > Properties: Resource: Text file encoding: Inherited from container (**Cp1252**)_
    2. Inside the new project create a new **class** `Äpp`<br>→ right-click `Äpp.java`: _Properties: Resource: Text file encoding: (inherited from container: **Cp1252**)_<br>→ **no** _"Project has no explicit encoding set"_ warning
4. With _Eclipse IDE for Java Developers **2022-06** RC2_:
    1. Create a new **workspace**<br>→ preferences _General > Workspace: Text file encoding: Other (**UTF-8**)_
    2. Inside the new workspace create a new **project** _Hello Wörld_<br>→ _Project > Properties: Resource: Text file encoding: Other (**UTF-8**)_
    3. Inside the new project create a new **class** `Äpp`<br>→ right-click `Äpp.java`: _Properties: Resource: Text file encoding: Default (inherited from container: **UTF-8**)_

**Comment by jonahgraham** [comment]
The last comment was with the EPP zips which include Java 17. I also ran it with Oomph installed Eclipse using my installed Java 11 and got the same results.

**Comment by howlger** [comment]
@jonahgraham and @iloveeclipse thanks a lot for taking your time to verify this!

It's caused by **Java 18**:
The unwanted behavior only occurs when [JustJ Adoptium OpenJDK Hotspot JRE Complete 18.0.1](https://download.eclipse.org/justj/jres/18/updates/release/18.0.1/index_super.html) is installed or when running Eclipse 2022-06 (4.24) with Java 18 installed on the system (tested with Temurin-18.0.1+10).

With the JustJ **Java 17**.0.3 with which all Eclipse 2022-06 IDE packages are shipped, **everything works fine**.

**Comment by iloveeclipse** [comment]
> The unwanted behavior only occurs when [JustJ Adoptium OpenJDK Hotspot JRE Complete 18.0.1](https://download.eclipse.org/justj/jres/18/updates/release/18.0.1/index_super.html) is installed or when running Eclipse 2022-06 (4.24) with Java 18 installed on the system (tested with Temurin-18.0.1+10).

Holger, can you please report dedicated bug here for this. I assume system encoding is not properly recognized anymore (and utf-8 detected instead), that should be fixed.

**Comment by jonahgraham** [comment]
I wonder if Java 18 understands the new Windows option of using UTF-8 "Use Unicode UTF-8 for worldwide language support" and that causes the encoding to change? See https://bugs.eclipse.org/bugs/show_bug.cgi?id=572294 for another example of that setting being problematic.

**Comment by iloveeclipse** [comment]
> I wonder if Java 18 understands the new Windows option of using UTF-8 "Use Unicode UTF-8 for worldwide language support" and that causes the encoding to change?

No, I guess the code just asks Java for default encoding and gets garbage instead. Haven't debugged this yet, but that was the point of  https://openjdk.java.net/jeps/400  :)

**Comment by Bananeweizen** [comment]
Closing, as the remaining discussion should probably happen in #154. Feel free to reopen if you disagree.

**Comment by iloveeclipse** [comment]
There are two different issues: wrong system encoding detection with Java 18 and "restore to default encoding" behavior in general.
